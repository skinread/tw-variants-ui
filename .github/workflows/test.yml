# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and run tests across different versions of node
# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs

name: CI Tests

on:
  pull_request:
    branches: ["main","develop"]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [20.x]
    steps:
    - name: Checkout
      uses: actions/checkout@v4
    - name: Setup Node ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        cache-dependency-path: "**/package-lock.json"
    - name: Cache or restore dependencies
      uses: actions/cache@v4
      with:
        path: ~/.npm
        key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
        restore-keys: ${{ runner.os }}-npm-
    - name: Install dependencies
      run: npm ci

    - name: Check formatting and linting
      run: npm run prettier && npm run lint

    - name: Check package build
      run: npm run build

    - name: Get Playwright version
      run: echo "PLAYWRIGHT_VERSION=$(jq -r '.devDependencies["@playwright/test"]' package.json)" >> $GITHUB_ENV
    - name: Cache or restore Playwright binaries
      uses: actions/cache@v4
      with:
        path: |
          ~/.cache/ms-playwright
        key: ${{ runner.os }}-playwright-${{ env.PLAYWRIGHT_VERSION }}
        restore-keys: ${{ runner.os }}-playwright-
    - name: Install Playwright
      if: steps.playwright-cache.outputs.cache-hit != 'true'
      run: npm run pretest
    - name: Storybook test build
      run: npm run storybook:build -- --test
    - name: Storybook test suite
      run: npm run test:ci
