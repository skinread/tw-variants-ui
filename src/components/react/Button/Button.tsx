import { buttonVariants, type ButtonVariantsWithDocs } from '@common';

type ButtonPropsFiltered = Omit<
  React.ComponentPropsWithoutRef<'button'>,
  keyof ButtonVariantsWithDocs
>;

export interface ButtonProps extends ButtonVariantsWithDocs, ButtonPropsFiltered {
  /**
   * React children passthrough for the button content, typically just a string
   */
  children: React.ReactNode;
  /**
   * Not recommended, className override does not yet merge with tailwind-variants as expected
   */
  className?: string;
}

/**
 * Styled button element. For navigation link buttons use Button Link. Also exposes intrinsic `<button>` attributes.
 */
export const Button = (props: ButtonProps) => {
  const { children, color = 'primary', fullWidth = false, ...elemAttrs } = props;

  return (
    <button className={buttonVariants({ color, fullWidth })} {...elemAttrs}>
      {children}
    </button>
  );
};
