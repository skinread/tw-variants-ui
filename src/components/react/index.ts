export * from './Autocomplete/Autocomplete';
export * from './Button/Button';
export * from './ButtonLink/ButtonLink';
export * from './Checkbox/Checkbox';
export * from './Feedback/Feedback';
export * from './Icon/Icon';
export * from './InputField/InputField';
export * from './Modal/Modal';
export * from './RadioButtons/RadioButtons';
export * from './Select/Select';
export * from './TextArea/TextArea';
/* generator:new component */
