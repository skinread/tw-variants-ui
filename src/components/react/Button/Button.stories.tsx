import type { Meta, StoryObj } from '@storybook/react';
import { within, userEvent, fn } from '@storybook/test';

import { Icon } from '@components/react';
import { Button } from './Button';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  title: 'React/Button',
  component: Button,
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  args: {
    children: 'Press here',
    color: 'primary',
    fullWidth: false,
    onClick: fn(),
  },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {};

export const Secondary: Story = {
  args: {
    color: 'secondary',
  },
};

export const WithIcon: Story = {
  args: {
    children: (
      <>
        <Icon name="info" width={19} className="mr-1" />
        <span>More info</span>
      </>
    ),
  },
};

export const FullWidth: Story = {
  args: {
    fullWidth: true,
  },
  parameters: {
    layout: 'padded',
  },
};

/**
 * Open this story directly to observe automated test
 */
export const InteractionTest: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    await userEvent.click(canvas.getByRole('button'));
  },
};
