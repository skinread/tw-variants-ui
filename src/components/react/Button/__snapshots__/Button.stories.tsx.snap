// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React/Button FullWidth smoke-test 1`] = `
<button class="btn no-animation text-base h-auto min-h-fit px-2 py-2 btn-primary focus:bg-primary-focus focus:border-primary-focus hover:bg-primary-focus disabled:opacity-40 disabled:bg-primary disabled:text-white btn-block">
  Press here
</button>
`;

exports[`React/Button InteractionTest play-test 1`] = `
<button class="btn no-animation text-base h-auto min-h-fit px-2 py-2 btn-primary focus:bg-primary-focus focus:border-primary-focus hover:bg-primary-focus disabled:opacity-40 disabled:bg-primary disabled:text-white">
  Press here
</button>
`;

exports[`React/Button Primary smoke-test 1`] = `
<button class="btn no-animation text-base h-auto min-h-fit px-2 py-2 btn-primary focus:bg-primary-focus focus:border-primary-focus hover:bg-primary-focus disabled:opacity-40 disabled:bg-primary disabled:text-white">
  Press here
</button>
`;

exports[`React/Button Secondary smoke-test 1`] = `
<button class="btn no-animation text-base h-auto min-h-fit px-2 py-2 btn-outline text-primary border-primary hover:bg-primary/10 hover:border-primary hover:text-primary focus:bg-primary/10 active:bg-primary/30 disabled:opacity-40 disabled:border-primary disabled:text-primary disabled:bg-white">
  Press here
</button>
`;

exports[`React/Button WithIcon smoke-test 1`] = `
<button class="btn no-animation text-base h-auto min-h-fit px-2 py-2 btn-primary focus:bg-primary-focus focus:border-primary-focus hover:bg-primary-focus disabled:opacity-40 disabled:bg-primary disabled:text-white">
  <svg xmlns="http://www.w3.org/2000/svg"
       viewbox="0 0 24 24"
       fill="currentColor"
       class="inline-block  mr-1 icon-info"
       width="19"
       aria-hidden="true"
  >
    <path d="M12,24A12,12,0,1,1,24,12,12.013,12.013,0,0,1,12,24ZM12,2A10,10,0,1,0,22,12,10.011,10.011,0,0,0,12,2Z"
          stroke="none"
    >
    </path>
    <path d="M14,19H12V12H10V10h2a2,2,0,0,1,2,2Z"
          stroke="none"
    >
    </path>
    <circle cx="12"
            cy="6.5"
            r="1.5"
    >
    </circle>
  </svg>
  <span>
    More info
  </span>
</button>
`;
